internal/controller/children.go

```diff
@@ -367,9 +370,70 @@ func (c ChildResourceUpdates) failureMessage(pod *v1.Pod) string {
 	if msg != "" {
 		return msg
 	}
+
 	return ""
 }
 
+type EventsInferedStateResult = int
+
+const (
+	EisrNone            = -1
+	EisrInitiallyFailed = 0
+	EisrTemporaryFailed = 1
+	EisrFinallyFailed   = 2
+	EisrAutoScheduling  = 3
+)
+
+// eventsInferedFailure looks into the events of the session pod to
+// figure out whether a failure exists. The events are looked at with
+// the latest first, if an event occurs that indicates a scale-up or a
+// scheduled reason, then it is considered non-failing. If the first
+// event encountered is a "FailedScheduling" reason, then it indicates
+// a failure iff this state has occurred for a pre-defined period.
+//
+// Returns one of:
+// - initially failed: if the latest event is FailedScheduling and none of these has been seen before
+// - temporary failed: if the latest event is FailedScheduling but the wait timeout has not been reached
+// - finally failed: if the latest event is FailedScheduling and this has been seen for a while exceeding maximum wait time
+// - auto scheduling: if an TriggeredScaleUp events has been found as the last event
+// - none of the above
+func EventsInferedState(cr *amaltheadevv1alpha1.AmaltheaSession, client client.Reader, ctx context.Context) (EventsInferedStateResult, error) {
```

Comment:
Can be referenced as a new issue to solve later: in all `go` code I have read, `ctx context.Context` is always the first argument to a function.